<!DOCTYPE html>
<html>
<body>

<h1>LncLOOM: Definitions of Statistical values</h1>

<p>LncLOOM determines the statistical significance empirically through a number of iterations (set by the user) on two sets of random sequences.
<br>
<br> Firstly, LncLOOM performs motif discovery on a random set of sequences that have the same percentage identity and the same lengths as the orginal input sequences.
<br>This set of sequences is generated by shuffling a MSA (calculated by MAFFT) of the input sequences.
<br> 
<br> Secondly, LncLOOM performs motif discovery on a random set of sequences that have the same dicnucleotide compositions and the same lengths as the orginal input sequences.
<br>
From the iterative motif discovery performed on these random sets, LncLOOM calculates four statistical values:
</p>

<h4>E(i): Probability of finding the exact motif, at the same depth, in a random set of sequences that have the same percentage identity as the input sequences</h4>

<h4>P(i): Probability of finding any combination of the same number of motifs of the same length, or longer, at the same depth, in a random set of sequences that have the same percentage identity as the input sequences</h4>

<h4>E(r): Probability of finding the exact motif, at the same depth, in a random set of sequences that have the dinucleotide composition as the input sequences</h4>

<h4>P(r): Probability of finding any combination of the same number of motifs of the same length, or longer, at the same depth, in a random set of sequences that have the same dinucleotide composition as the input sequences</h4>

<br><br>
<h2>Troubleshooting statistical iterations</h2>
<h4>E(i) and P(i) values are calculated as "Undefined"</h4>
<p>The E(i) and P(i) values are calculated from a shuffled MSA that is produced by MAFFT, during the LncLOOM run.
<br>If LncLOOM, or Python, is not able to run MAFFT then these values will be "Undefined":
<br>1) If you are running on a server, ensure that the MAFFT module has been loaded.<br>
2) Make sure there are no mistakes in your fasta file of input sequences. For example: MAFFT may fail if there are unrecognizable letters. <br>
3) If the fasta file is very large or sequences are very long, it is possible that more memory is needed to calculate the MSA</p>

</body>
</html>
